### PR TITLE
CXX-751 - Specify timeouts with chrono

### PR DESCRIPTION
**Pull request by tessavitabile**
I had a couple questions:

For the find options, why do we have both max_time and max_await_time? max_time is not used.

For the methods where the max_time is not used, should I investigate whether it's possible to pass it to the C driver yet?

### CODE REVIEW

File: src/mongocxx/options/aggregate.cpp [hunk 1]
```diff
@@ -25,8 +25,8 @@ void aggregate::allow_disk_use(bool allow_disk_use) {
 void aggregate::batch_size(std::int32_t batch_size) {
     _batch_size = batch_size;
 }
-void aggregate::max_time_ms(std::int64_t max_time_ms) {
-    _max_time_ms = max_time_ms;
+void aggregate::max_time(std::chrono::milliseconds max_time) {
+    _max_time = max_time;
```
Comment (acmorrow) [inline src/mongocxx/options/aggregate.cpp]: nit: std::move(max_time)

File: src/mongocxx/options/aggregate.cpp [hunk 2]
```diff
@@ -44,8 +44,8 @@ const stdx::optional<bool>& aggregate::allow_disk_use() const {
 const stdx::optional<std::int32_t>& aggregate::batch_size() const {
     return _batch_size;
 }
-const stdx::optional<std::int64_t>& aggregate::max_time_ms() const {
-    return _max_time_ms;
+const stdx::optional<std::chrono::milliseconds>& aggregate::max_time() const {
```
Comment (acmorrow) [inline src/mongocxx/options/aggregate.cpp]: can you add a blank line before this method, and after

File: src/mongocxx/options/find.cpp [hunk 3]
```diff
@@ -42,12 +42,12 @@ void find::limit(std::int32_t limit) {
     _limit = limit;
 }
 
-void find::max_await_time_ms(std::int64_t max_await_time_ms) {
-    _max_await_time_ms = max_await_time_ms;
+void find::max_await_time(std::chrono::milliseconds max_await_time) {
+    _max_await_time = max_await_time;
```
Comment (acmorrow) [inline src/mongocxx/options/find.cpp]: std::move, here and below. I'll stop mentioning it here, but please check the other setters

**Comment by acmorrow** [comment]
@tessavitabile - looks like a good start. Two your two questions: yes, please investigate both.

**Comment by acmorrow** [comment]
@tessavitabile - I'd recommend checking in with @ajdavis on these.

**Comment by tessavitabile** [comment]
I emailed Jesse about those questions.

File: src/mongocxx/options/aggregate.cpp (hunk 2 above)
Comment (tessavitabile) [inline src/mongocxx/options/aggregate.cpp]: There are blank lines missing between all methods, so I'll clang-format before I push.

**Comment by tessavitabile** [comment]
@ajdavis - Would you be willing to check the last commit to see whether I'm passing maxTimeMS correctly?

File: src/mongocxx/collection.cpp [hunk 4]
```diff
@@ -426,6 +430,8 @@ stdx::optional<bsoncxx::document::value> collection::find_one_and_replace(
         flags |= ::MONGOC_FIND_AND_MODIFY_RETURN_NEW;
     }
 
+    // TODO: use options.max_time() when available in the C driver
```
Comment (acmorrow) [inline src/mongocxx/collection.cpp]: Please file a new CXX driver ticket in the C++11 driver epic, and set the fix version to "waiting on dependency", and add the "waiting-on-mongoc" label. The ticket should be linked as "depends on" the relevant C driver ticket. Also, how should we note this in the header comments? Or should this be a relnote?

**Comment by acmorrow** [comment]
This LGTM, with a request to do some project management work around the feature we can't implement yet.

File: src/mongocxx/collection.cpp (hunk 4 above)
Comment (tessavitabile) [inline src/mongocxx/collection.cpp]: Will do. Sorry, what is a relnote?

Comment (acmorrow) [inline src/mongocxx/collection.cpp]: Sorry for the jargon. Release note. I'd like to add this as a "known issue". We don't really have a plan for how to do that, yet.

**Comment by acmorrow** [comment]
@tessavitabile  - It looks like some of the travis builders are failing

**Comment by tessavitabile** [comment]
@ajdavis gave this the LGTM, so I'll just look into the travis builder failing.

File: src/mongocxx/collection.cpp [hunk 5]
```diff
@@ -185,6 +186,10 @@ cursor collection::find(bsoncxx::document::view filter, const options::find& opt
         filter_builder << "$orderby" << bsoncxx::types::b_document{*options.sort()};
     }
 
+    if (options.max_time()) {
+        filter_builder << "$maxTimeMS" << (*options.max_time()).count();
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: optional will let you call methods on the underlying object directly with -> so:
```
options.max_time()->count();
```

File: src/mongocxx/collection.cpp [hunk 6]
```diff
@@ -198,8 +203,8 @@ cursor collection::find(bsoncxx::document::view filter, const options::find& opt
         options.limit().value_or(0), options.batch_size().value_or(0), filter_bson.bson(),
         projection.bson(), rp_ptr);
 
-    if (options.max_await_time_ms()) {
-        libmongoc::cursor_set_max_await_time_ms(mongoc_cursor, *options.max_await_time_ms());
+    if (options.max_await_time()) {
+        libmongoc::cursor_set_max_await_time_ms(mongoc_cursor, (*options.max_await_time()).count());
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: Also -> would be clearer here

File: src/mongocxx/collection.cpp [hunk 7]
```diff
@@ -238,8 +243,8 @@ cursor collection::aggregate(const pipeline& pipeline, const options::aggregate&
         inner << close_document;
     }
 
-    if (options.max_time_ms()) {
-        b << "maxTimeMS" << *options.max_time_ms();
+    if (options.max_time()) {
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: can you combine this with the above conditional on options.max_time() ?

File: src/mongocxx/collection.cpp [hunk 8]
```diff
@@ -632,10 +647,14 @@ bsoncxx::document::value collection::create_index(bsoncxx::document::view keys,
...
                             const options::distinct& options) {
-    auto command = bsoncxx::builder::stream::document{}
-                   << "distinct" << name() << "key" << field_name << "query"
-                   << bsoncxx::types::b_document{query} << bsoncxx::builder::stream::finalize;
-    scoped_bson_t command_bson{command};
+    bsoncxx::builder::stream::document command{};
+    command << "distinct" << name() << "key" << field_name << "query" << bsoncxx::types::b_document{query};
+
+    if (options.max_time()) {
+        command << "maxTimeMS" << (*options.max_time()).count();
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: operator ->

File: src/mongocxx/collection.cpp (hunk 7 above)
Comment (tessavitabile) [inline src/mongocxx/collection.cpp]: Sorry, I don't know what conditional you're referring to. Could you clarify?

File: src/mongocxx/options/aggregate.hpp [hunk 9]
```diff
@@ -17,6 +17,7 @@
 #include <mongocxx/config/prelude.hpp>
 
 #include <cstdint>
+#include <chrono>
```
Comment (samantharitter) [inline src/mongocxx/options/aggregate.hpp]: please put above <cstdint>

File: src/mongocxx/options/count.hpp [hunk 10]
```diff
@@ -18,6 +18,7 @@
 
 #include <cstdint>
 #include <string>
+#include <chrono>
```
Comment (samantharitter) [inline src/mongocxx/options/count.hpp]: alphabetize

File: src/mongocxx/collection.cpp (hunk 7 above)
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: oh sorry I didn't realize this was a different method!

File: src/mongocxx/collection.cpp (hunk 5 above)
Comment (acmorrow) [inline src/mongocxx/collection.cpp]: +1

File: src/mongocxx/collection.cpp (hunk 6 above)
Comment (acmorrow) [inline src/mongocxx/collection.cpp]: I think there are several other places too. @tessavitabile please change them all.

File: src/bsoncxx/document/value.cpp [hunk 11]
```diff
@@ -41,6 +41,7 @@ void uint8_t_deleter(std::uint8_t* ptr) {
 value::value(document::view view)
     : _data(new std::uint8_t[static_cast<std::size_t>(view.length())], uint8_t_deleter),
       _length(view.length()) {
+    std::cout << "length " << view.length() << " data " << view.data() << std::endl;
```
Comment (acmorrow) [inline src/bsoncxx/document/value.cpp]: Looks unrelated

File: src/mongocxx/collection.cpp [hunk 12]
```diff
@@ -17,6 +17,7 @@
 #include <cstdint>
 #include <tuple>
 #include <utility>
+#include <chrono>
```
Comment (acmorrow) [inline src/mongocxx/collection.cpp]: alphabetize

File: src/mongocxx/options/aggregate.hpp (hunk 9 above)
Comment (acmorrow) [inline src/mongocxx/options/aggregate.hpp]: +1

File: src/mongocxx/options/distinct.hpp [hunk 13]
```diff
@@ -18,6 +18,7 @@
 
 #include <cstdint>
 #include <string>
+#include <chrono>
```
Comment (acmorrow) [inline src/mongocxx/options/distinct.hpp]: here too for alphabetize

File: src/mongocxx/options/find.hpp [hunk 14]
```diff
@@ -17,6 +17,7 @@
 #include <mongocxx/config/prelude.hpp>
 
 #include <cstdint>
+#include <chrono>
```
Comment (acmorrow) [inline src/mongocxx/options/find.hpp]: here too, I'll stop mentioning it now

File: src/mongocxx/test/collection.cpp [hunk 15]
```diff
@@ -534,21 +534,41 @@ TEST_CASE("CRUD functionality", "[driver::collection]") {
 
         // copy into a vector.
         std::vector<bsoncxx::document::value> results;
+        try {
```
Comment (acmorrow) [inline src/mongocxx/test/collection.cpp]: changes in this file look unrelated.

**Comment by acmorrow** [comment]
Overall this LGTM, mod the following systemic issues:
- ordering of the `<chrono>` header
- switching to using `operator->` on known-engaged optionals
- Fixing up some straggler code that looks like it was unrelated test code.

**Comment by tessavitabile** [comment]
Sorry about that, I'm still trying to figure out why travis is failing.

File: src/mongocxx/test/collection.cpp [hunk 16]
```diff
@@ -534,21 +534,25 @@ TEST_CASE("CRUD functionality", "[driver::collection]") {
 
         // copy into a vector.
         std::vector<bsoncxx::document::value> results;
+        
         for (auto&& result : distinct_results) {
```
Comment (tessavitabile) [inline src/mongocxx/test/collection.cpp]: An exception is being thrown at this line on travis. I didn't change the output of collection::distinct, and I'm not able to reproduce the exception locally. Does anyone have an idea of why this might be occurring?

Comment (acmorrow) [inline src/mongocxx/test/collection.cpp]: @tessavitabile - Yeah and the logs don't show much. If you can't repro locally, I'm inclined to move forward with the patch. I don't see how they are connected.

Comment (tessavitabile) [inline src/mongocxx/test/collection.cpp]: Okay, sounds good.
